src/hausdorff.jl

```diff
@@ -82,35 +82,44 @@ See also: [`hausdorff`](@ref)
 const ModifiedHausdorff = GenericHausdorff{MeanReduction, MaxReduction}
 ModifiedHausdorff() = ModifiedHausdorff(MeanReduction(), MaxReduction())
 
-# convert binary image to a point set format
-function img2pset(img::AbstractArray{T}) where T<:Union{Gray{Bool}, Bool}
-    inds = findall(x->x==true, img)
-    [inds[j][i] for i=1:ndims(img), j=1:length(inds)]
+# convert binary image to its distance transform
+const hausdorff_transform(img::AbstractArray{Bool}) = img |> feature_transform |> distance_transform
+
+const hausdorff_transform(img::GenericGrayImage) = hausdorff_transform(of_eltype(Bool, img))
+
+function evaluate_hausdorff(d::GenericHausdorff,
+                            imgA::AbstractArray{Bool},
+                            imgB::AbstractArray{Bool},
+                            dtA::AbstractArray,
+                            dtB::AbstractArray)
+    # trivial cases
+    imgA == imgB && return 0.
+    (isempty(imgA) || isempty(imgB)) && return Inf
+
+    # dtA and dtB contain the distance from each pixel
+    # to the nearest active pixel in imgA and imgB, respectively.
+    # We only care about the distances from imgA to imgB and vice
+    # versa, so we'll pull those out by using those images as logical
+    # masks into the distance arrays.
+    dAB = _reduce(d.inner_op, dtB[imgA])
+    dBA = _reduce(d.inner_op, dtA[imgB])
+    _reduce(d.outer_op, (dAB, dBA))
 end
-function img2pset(img::GenericGrayImage)
+function evaluate_hausdorff(d::GenericHausdorff,
+                            imgA::GenericGrayImage,
+                            imgB::GenericGrayImage,
+                            dtA::AbstractArray,
+                            dtB::AbstractArray)
     try
-        return img2pset(of_eltype(Bool, img))
+        evaluate_hausdorff(d, of_eltype(Bool, imgA), of_eltype(Bool, imgB), dtA, dtB)
     catch e
         e isa InexactError && throw(ArgumentError("Binary image is needed."))
         rethrow(e)
     end
 end
```

Comment:
Is this code outdated?

Reply:
It should probably be two lines as well like:

```julia
evaluate_hausdorff(d::GenericHausdorff, imgA::GenericGrayImage, ...) =
  evaluate_hausdorff(d::GenericHausdorff, of_eltype(Bool, imgA), ...)
```

Reply:
no, that's from the latest commit. Previously that line had multi-line formatting.

Reply:
Actually, I think it is not needed right? We are only calling the low level version with images of type AbstractArray{Bool}

Reply:
hmm well it depends whether you want the nice error message. Actually the same goes for the change to the `hausdorff_transform` function: previously with the try/catch you would get an error message saying "Binary image is needed" if you tried to pass a non-binary image. Now you get the following error:
```
julia> hausdorff([0.5 0.0; 0.0 0.5], [0.5 0.5; 0.0 0.5])
ERROR: InexactError: Bool(0.5)      
Stacktrace:
 [1] Bool at .\float.jl:73 [inlined]
 [2] convert at .\number.jl:7 [inlined]
 [3] #2 at C:\Users\John\.julia\packages\MappedArrays\4939e\src\MappedArrays.jl:116 [inlined]
 [4] mappedarray at C:\Users\John\.julia\packages\MappedArrays\4939e\src\MappedArrays.jl:89 [inlined]   
 [5] of_eltype at C:\Users\John\.julia\packages\MappedArrays\4939e\src\MappedArrays.jl:116 [inlined]    
 [6] hausdorff_transform(::Array{Float64,2}) at C:\Users\John\.julia\dev\ImageDistances\src\hausdorff.jl:88
 [7] GenericHausdorff at C:\Users\John\.julia\dev\ImageDistances\src\hausdorff.jl:121 [inlined]
 [8] hausdorff(::Array{Float64,2}, ::Array{Float64,2}) at C:\Users\John\.julia\dev\ImageDistances\src\hausdorff.jl:126
 [9] top-level scope at none:1
```

Reply:
My personal opinion is that the error message above is enough, but others may think differently. The code with try/catch is so verbose, we let other people comment if they prefer the more explicit message. cc: @johnnychen94 what do you think?

Reply:
Images with type `AbstractArray{Bool}` are handled in lines above. This method permits non `AbstractArray{Bool}` image type, e.g., outputs from `ImageBinarization.binarize`.

There might be better solutions than the explicit `try-catch` solution, but this doesn't belong to the scope of this PR.

Reply:
Nice @johnnychen94 , so we could stick to the one-liner for now and think about this error message in a separate PR if needed right?

Reply:
I would personally stick to whatever it was and wouldn't botter this (whether it should be oneliner or try-catch) at all in this PR.

I don't see any problem with the verbosity here. It's not always easy for new users to realize that `Hausdorff` distances only accepts `AbstractArray{Bool}` especially in the context that most other distances in this package allows generic arrays. In this case, `InexactError(0.5)` might be a source of confusion.

Reply:
I think to have the same behaviour as before (i.e., throwing the "Binary image is needed" error) I'll have to add another try-catch in `hausdorff_transform(::GenericGrayImage)`. @juliohm does that make sense to you?

Reply:
Makes sense, I am fine with either approach. It seems reasonable to move the error handling to the top level before doing any calculation. It should happen in a single place before the image conversions. It is thrown by the `to_eltype(Bool, img)` right? So the function calling this one should handle the error.

Reply:
@juliohm take a look at the latest commit. I added the error handling to both `hausdorff_transform` and `evaluate_hasdorff`, because they both call `to_eltype`.